src/execution_plan/optimizations/optimize_cartesian_product.c

```diff
@@ -7,115 +7,114 @@
+	return solving_branches;
 }
 
 static void _optimize_cartesian_product(ExecutionPlan *plan, OpBase *cp) {
 	// Retrieve all filter operations located upstream from the Cartesian Product.
-	OpFilter **filter_ops = _locate_filters(cp);
-	uint filter_count = array_len(filter_ops);
-
-	// For each stream joined by the Cartesian product, collect all entities the stream resolves.
-	int stream_count = cp->childCount;
-	rax *stream_entities[stream_count];
-	OptimizeUtils_BuildStreamFromOp(cp, stream_entities, stream_count);
+	Filter_And_Entities *filters = _locate_filters_and_entities(cp);
+	uint filter_count = array_len(filters);
 
 	for(uint i = 0; i < filter_count; i ++) {
-		// Try to create a new dual-branched cartesian product, followed by the current filter.
-		OpFilter *filter_op = filter_ops[i];
-
-		/* A new cartesian product can be generated if both sides of the filter can be fully and
-		 * separately resolved by exactly two child streams. */
-		FT_FilterNode *f = filter_op->filterTree;
-
-		/* Make sure LHS of the filter is resolved by a stream. */
-		AR_ExpNode *lhs = f->pred.lhs;
-		uint lhs_resolving_stream = OptimizeUtils_RelateExpToStream(lhs, stream_entities, stream_count);
-		if(lhs_resolving_stream == NOT_RESOLVED) continue;
-		/* Make sure RHS of the filter is resolved by a stream. */
-		AR_ExpNode *rhs = f->pred.rhs;
-		uint rhs_resolving_stream = OptimizeUtils_RelateExpToStream(rhs, stream_entities, stream_count);
-		if(rhs_resolving_stream == NOT_RESOLVED) continue;
+		// Try to create a cartesian product, followed by the current filter.
+		OpFilter *filter_op = filters[i].filter;
+		ExecutionPlan_RemoveOp(plan, (OpBase *)filter_op);
 
-		// This stream is solved by a single cartesian product child and needs to be propagated up.
-		if(lhs_resolving_stream == rhs_resolving_stream) {
-			OptimizeUtils_MigrateFilterOp(plan, cp->children[rhs_resolving_stream], filter_op);
+		OpBase **solving_branches = _find_entities_solving_branches(filters[i].entities, cp);
+		// This filter is solved by a single cartesian product child and needs to be propagated up.
+		if(array_len(solving_branches) == 1) {
+			ExecutionPlan_PushBelow(solving_branches[0], (OpBase *)filter_op);
+			array_free(solving_branches);
 			continue;
 		}
 
-		// Retrieve the relevant branch roots.
-		OpBase *right_branch = cp->children[rhs_resolving_stream];
-		OpBase *left_branch = cp->children[lhs_resolving_stream];
-		// Detach the streams and completely remove the filter from the execution plan.
-		ExecutionPlan_DetachOp(right_branch);
-		ExecutionPlan_DetachOp(left_branch);
-		ExecutionPlan_RemoveOp(plan, (OpBase *)filter_op);
-		// Combine the streams as cartesian product and filter them.
-		OpBase *filtered_cp_branch = _chain_cp_and_filter(cp->plan, left_branch, right_branch,
-														  (OpBase *)filter_op);
+		// Need to create a new cartesian product and connect the solving branches to the filter.
+		OpBase *new_cp = NewCartesianProductOp(cp->plan);
```

Comment:
Can't we use given `plan` argument instead of `cp->plan` ?

Reply:
no, it messes the record map